src/embedded_flash/helper.rs

```diff
@@ -0,0 +1,282 @@
+/// All objects with length of 0, regardless of the start value, are considered empty.
+pub struct ModRange {
+    start: usize,
+    length: usize,
+}
+
+impl ModRange {
+    /// Returns a new range of given start and length.
+    ///
+    /// The stored length can be shorter than what is passed in, so that the range does not include
+    /// values bigger than `usize::MAX`.
+    pub fn new(start: usize, mut length: usize) -> ModRange {
+        if length > 0 {
+            length = start.saturating_add(length - 1) - start + 1;
+        }
+        ModRange { start, length }
+    }
+
+    /// Create a new empty range.
+    pub fn new_empty() -> ModRange {
+        ModRange::new(0, 0)
+    }
+
+    /// Returns the start of the range.
+    pub fn start(&self) -> usize {
+        self.start
+    }
+
+    /// Returns the length of the range.
+    pub fn length(&self) -> usize {
+        self.length
+    }
+
+    /// Returns whether this range contains any addresses.
+    pub fn is_empty(&self) -> bool {
+        self.length == 0
+    }
+
+    /// Returns the disjoint union with the other range, if is consecutive.
+    ///
+    /// Appending empty ranges is not possible.
+    /// Appending to the empty range returns the other range.
+    pub fn append(&self, other: ModRange) -> Option<ModRange> {
+        if self.is_empty() {
+            return Some(other);
+        }
+        if other.is_empty() {
+            return None;
+        }
+        if self.start >= other.start {
+            return None;
+        }
+        if self.length != other.start - self.start {
+            return None;
+        }
+        let new_length = self.length.checked_add(other.length);
+        new_length.map(|l| ModRange::new(self.start, l))
+    }
+
+    /// Returns whether the given address is inside the range.
+    pub fn contains(&self, x: usize) -> bool {
+        // We want to check the 2 following inequalities:
+        // (1) `start <= x`
+        // (2) `x < start + length`
+        // However, the second one may overflow written as is. Using (1), we rewrite to:
+        // (3) `x - start <= length`
+        self.start <= x && x - self.start < self.length
+    }
+
+    /// Returns whether the given range is fully contained, i.e. their cut equals the parameter.
+    pub fn contains_range(&self, range: &ModRange) -> bool {
+        if range.is_empty() {
+            return true;
+        }
+        let end = range.start + (range.length - 1);
+        self.contains(range.start) && self.contains(end)
+    }
+
+    /// Returns an iterator for all contained numbers that are divisible by the modulus.
+    pub fn aligned_iter(&self, modulus: usize) -> impl Iterator<Item = usize> {
+        let remainder = self.start % modulus;
+        // Saturating first and limit in case they are usize::MAX.
+        let first = if remainder == 0 {
+            self.start
+        } else {
+            (self.start - remainder).saturating_add(modulus)
```

Comment:
This should be clarified by my above comment? I basically don't care about `usize::MAX`. And if the addition saturation, the returned iterator is empty, so it can't return anything that is not congruent to 0.

Reply:
Yes, I see why it's not a problem if MAX is not a possible value.

Reply:
Should be documented!